core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -666,7 +671,7 @@ else if (scaleWriters
                     // do not insert an exchange if partitioning is compatible
                     && !newSource.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, plannerContext.getMetadata(), session)) {
                 if (partitioningScheme.get().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
-                    partitioningScheme = Optional.of(partitioningScheme.get().withPartitioningHandle(SCALED_WRITER_HASH_DISTRIBUTION));
+                    partitioningScheme = Optional.of(partitioningScheme.get().withPartitioningHandle(SCALED_WRITER_HASH_DISTRIBUTION).withPartitionCount(maxWritersNodesCount));
```

Comment:
What about when scaleWriters is disabled? In that case also we need to set partitionCount

Reply:
Then `partitionCount` is set in a such case. Please look at: https://github.com/trinodb/trino/pull/16238/files/6906e1162586ca4385a9e65b4ba1669b264ed130#diff-12f27276ffd16cf70db9d0858401083afe01f9bb796892b2b6e81803430954c2L625

It must be set there because it is how `PlanFragmenter` and `ApplyPreferedPartitioning` works together - a bit complicated.